File: src/main/java/com/baulsupp/oksocial/network/DnsOverride.java
package com.baulsupp.oksocial.network;

import com.google.common.collect.Maps;
import ee.schimke.oksocial.output.util.UsageException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.logging.Logger;
import okhttp3.Dns;

public class DnsOverride implements Dns {
  private static Logger logger = Logger.getLogger(DnsOverride.class.getName());

  private final Dns dns;
  private Map<String, String> overrides = Maps.newHashMap();

  public DnsOverride(Dns dns) {
    this.dns = dns;
  }

  private void put(String host, String target) {
    overrides.put(host, target);
  }

  @Override public List<InetAddress> lookup(String hostname) throws UnknownHostException {
    String override = overrides.get(hostname);

    if (override != null) {
      logger.fine("Using Dns Override (" + hostname + "): " + override);
      return Collections.singletonList(InetAddress.getByName(override));
    }

    return dns.lookup(hostname);
  }

  public static DnsOverride build(Dns dns, List<String> resolveStrings) {
    DnsOverride dnsOverride = new DnsOverride(dns);

    for (String resolveString : resolveStrings) {
      String[] parts = resolveString.split(":");

      if (parts.length != 2) {
        throw new UsageException("Invalid resolve string '" + resolveString + "'");
      }

      dnsOverride.put(parts[0], parts[1]);
    }

    return dnsOverride;
  }
}

File: src/main/java/com/baulsupp/oksocial/network/DnsOverride.kt
package com.baulsupp.oksocial.network

import com.baulsupp.oksocial.output.util.UsageException
import com.google.common.collect.Maps
import okhttp3.Dns
import java.net.InetAddress
import java.net.UnknownHostException
import java.util.logging.Logger

class DnsOverride(private val dns: Dns) : Dns {
  private val overrides = Maps.newHashMap<String, String>()

  private fun put(host: String, target: String) {
    overrides.put(host, target)
  }

  @Throws(UnknownHostException::class)
  override fun lookup(hostname: String): List<InetAddress> {
    val override = overrides[hostname]

    if (override != null) {
      logger.fine("Using Dns Override ($hostname): $override")
      return listOf(InetAddress.getByName(override))
    }

    return dns.lookup(hostname)
  }

  companion object {
    private val logger = Logger.getLogger(DnsOverride::class.java.name)

    fun build(dns: Dns, resolveStrings: List<String>): DnsOverride {
      val dnsOverride = DnsOverride(dns)

      for (resolveString in resolveStrings) {
        val parts = resolveString.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        if (parts.size != 2) {
          throw UsageException("Invalid resolve string '$resolveString'")
        }

        dnsOverride.put(parts[0], parts[1])
      }

      return dnsOverride
    }
  }
}
